mvebu: add support for Check Point V-80

musashino205 opened the pull request:
This patch series adds support for Check Point V-80 appliance.

musashino205 committed:
mvebu: add specific signature support to boot-scr and sdcard-img-ext4

Add image-specific signature support to Build/boot-scr and
Build/sdcard-img-ext4.
These are required to switch root device passed to the kernel on Check
Point V-80.

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
mvebu: add support for Check Point V-80

Check Point V-80 (Quantum Spark 1550 Appliance) is an Appliance, based
on Armada 7040 (88F7040).

Specification:

- SoC             : Marvell Armada 7040 (88F7040)
- RAM             : DDR4 2 GiB (4x Nanya NT5AD512M8D3-HR)
- Flash           : eMMC 4 GiB (Toshiba THGBMNG5D1LBAIL)
- Ethernet        : 6x 10/100/1000 Mbps
  - LAN 1-5       : Marvell 88E6352
  - WAN           : Marvell 88E1512
- LEDs/Keys (GPIO): 6x/1x
- UART            : "CONSOLE" port (USB 1.1 Type-C)
  - chip          : Silicon Labs CP210x
  - port          : ttyS0
  - settings      : 115200bps 8n1
- HW Monitoring   : 2x nuvoTon NCT7802Y
- USB             : USB 3.0 Type-A
- Power           : 12 VDC, 3.3 A
  - plug          : DC Plug 2.5/5.5 mm (inner/outer)

Flash instruction (common):

1. Boot V-80 normally
2. Login to the vendor CLI (default: admin/admin) and login to the Linux
   CLI by `expert` command
3. Update U-Boot environment variables by the following commands

   fw_setenv bootcmd_ow_usb 'usb start; load usb 0:1 ${loadaddr} boot.scr && source ${loadaddr}'
   fw_setenv bootcmd_ow_emmc 'run set_mmc_internal; mmc read ${loadaddr} ${prim_header_mmc_blk} 4 && source ${loadaddr}'
   fw_setenv bootcmd 'set cp_quiet; run set_bootargs_vx; run bootcmd_ow_usb; run bootcmd_ow_emmc; run bootcmd_part${activePartition};'

   Attention: don't forget single quatations of values to prevent
              expansion of variables

4. Turn off the device

Flash instruction (USB-boot):

1. Burn (squashfs|ext4)-usb.img.gz to USB storage
2. Connect that storage to the USB 3.0 port on V-80
3. Turn on V-80 and it will be booted with OpenWrt in that USB storage

Flash instruction (eMMC-boot):

1. Copy initramfs image, dtb and bootsctipt to the USB storage with
   renaming

   initramfs.bin -------> Image
   dtb -----------------> armada-7040-v-80.dtb
   bootscript (.scr) ---> boot.scr

2. Connect that storage to the USB 3.0 port on V-80
3. Turn on V-80 and it will be booted with OpenWrt initramfs image in
   that USB storage
4. Upload (squashfs|ext4)-sysupgrade.gz to V-80
5. Perform sysupgrade with the uploaded image
6. Wait ~100 seconds to complete flashing

Reverting to stock firmware:

1. Turn on V-80 and interrupt booting by Ctrl + C
2. Select "4. Restore to Factory Defaults (local)"
3. Wait ~180 seconds to complete restoring and rebooting

Notes:

- V-80 has some HW versions. Internal MicroSD card slot and mPCIe slot
  is available on some HW versions, but not on the other HW versions.

  confirmed (MicroSD/mPCIe slots):

  - 1.0.1: unavailable
  - 1.0.3: available

- The partition table in the internal eMMC has single partition, but
  "blkdevparts=" parameter will be passed from the bootloader and that
  definition will be used instead.

MAC addresses:

LAN: 00:1C:7F:xx:xx:4B (mmcblk1boot0, ethaddr (text))
WAN: 00:1C:7F:xx:xx:4A (mmcblk1boot0, eth1addr (text))

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
mvebu: cortexa72: enable CONFIG_CMDLINE_PARTITION for blkdevparts

Enable CONFIG_CMDLINE_PARTITION symbol to use "blkdevparts=" parameter
support on Check Point V-80.

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
uboot-envtools: add support for Check Point V-80

Add support for Check Point V-80 to add/edit bootcmd variables for
booting OpenWrt.

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
build: add dtb support for sysupgrade-tar

Add dtb support for Build/sysupgrade-tar definition and
sysupgrade-tar.sh script.
This changes are required for updating dtb separately.

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
base-files: add dtb support to emmc_upgrade_tar()

Add dtb support for emmc_upgrade_tar() to update dtb separately.

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
mvebu: cortexa72: enable emmc feature

Enable "emmc" feature to use the emmc helpers on sysupgrade.

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
base-files: return error when invalid partition table found in get_partitions()

Return error state instead of "exit" when invalid partition table is found
on the specified device in get_partitions().
Currently, this function exits without error state and firmware validation
by sysupgrade process will be passed if the invalid firmware image is
specified.

example (sysupgrade.gz instead of sdcard/usb.img.gz):

root@OpenWrt:/# sysupgrade /tmp/openwrt-mvebu-cortexa72-checkpoint_v-80-squashfs
-sysupgrade.gz
Tue Apr  9 04:45:52 UTC 2024 upgrade: Reading partition table from bootdisk...
Tue Apr  9 04:45:52 UTC 2024 upgrade: Extract boot sector from the image
Tue Apr  9 04:45:52 UTC 2024 upgrade: Reading partition table from image...
Tue Apr  9 04:45:52 UTC 2024 upgrade: Invalid partition table on /tmp/image.bs
Failed to parse message data
sh: out of range
Tue Apr  9 04:45:52 UTC 2024 upgrade: Saving config files...
Tue Apr  9 04:45:53 UTC 2024 upgrade: Commencing upgrade. Closing all shell sessions.
{
        "error": {
                "message": "Firmware image couldn't be validated: no JSON input"
        }
}
Command failed: Unknown error
root@OpenWrt:/#

After this fix:

root@OpenWrt:/# sysupgrade /tmp/openwrt-mvebu-cortexa72-checkpoint_v-80-squashfs
-sysupgrade.gz
Tue Apr  9 04:53:52 UTC 2024 upgrade: Reading partition table from bootdisk...
Tue Apr  9 04:53:52 UTC 2024 upgrade: Extract boot sector from the image
Tue Apr  9 04:53:52 UTC 2024 upgrade: Reading partition table from image...
Tue Apr  9 04:53:52 UTC 2024 upgrade: Invalid partition table on /tmp/image.bs
Image check failed.
root@OpenWrt:/#

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
treewide: handle an error returned from get_partitions()

Return error state to make the firmware validation fail when get_partition
fails, to prevent execution of sysupgrade with a invalid firmware.

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 commented:
UPDATE:

- rebase from current main branch

hauke commented on package/base-files/files/lib/upgrade/common.sh:
```diff
@@ -277,7 +277,7 @@ get_partitions() { # <device> <filename>
 		if [ "$magic" != $'\x55\xAA' ]; then
 			v "Invalid partition table on $disk"
-			exit
+			return 1
```
none of the callers is checking the return value of this function. Will we still abort in case of an error?

hauke commented on package/base-files/files/lib/upgrade/legacy-sdcard.sh:
```diff
@@ -7,12 +7,12 @@ legacy_sdcard_check_image() {
 	}
 
-	get_partitions "/dev/$diskdev" bootdisk
+	get_partitions "/dev/$diskdev" bootdisk || return 1
```
There are many other callers of `get_partitions`

musashino205 replied:
Oops, I had completely forgotten about that... I'll try to update other callers.

musashino205 committed:
base-files: add gzip-compressed tar support to emmc_upgrade_tar()

Add support for the tar archive compressed by gzip to emmc_upgrade_tar()
function.

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 commented:
UPDATE:

- handle an error returned from `get_partitions()` on all callers
- add gzip-compressed tar support to `emmc_upgrade_tar()`
  - remove unnecessary gzip decompression from platform.sh for V-80
- rebase from current main branch

musashino205 commented:
UPDATE:

- rebase from current main branch

musashino205 commented:
UPDATE:

- remove unused `vbus-supply` property from USB node
- add missing `bus-width = <4>;` into `&cp0_sdhci0 {}`
- rebase from current main branch

musashino205 commented:
UPDATE:

- rebase from current main branch
  - resolve a conflict in emmc.sh

musashino205 commented:
I'll close this PR, successor is #16904.
